### PR TITLE
Conflict resolution

### PR DESCRIPTION
**Pull request by bhousel**
This implements Conflict Resolution as discussed on #1053 

* At save time, retrieve latest versions of all modified nodes
* Merge non-conflicting changes to tags
* Merge non-conflicting changes to way nodelist (using diff3)
* ~~If a modified entity has been deleted remotely, restore it (related #2475)~~

Still to do
- [X] some minimal UI for listing the unmergable conflicts for the user
- [X] allow user to download the changeset if automatic merge fails
- [ ] merge non-conflicting changes to relation members for some types? (I think the `route` type would be safe to diff3)

@jfirebaugh, @aaronlidman want to take a look?

### CODE REVIEW

File: js/id/core/graph.js [hunk 1]
```diff
@@ -16,10 +16,7 @@ iD.Graph = function(other, mutable) {
 
     this.transients = {};
     this._childNodes = {};
-
-    if (!mutable) {
-        this.freeze();
-    }
+    this.frozen = !mutable;
```
Comment (jfirebaugh) [inline js/id/core/graph.js]: What's the rationale for this change?

File: js/id/actions/merge_remote_changes.js [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        }
+        else if (target.type === 'relation') {
+            target = mergeRemoteMembers(target);
+        }
+
+        target = mergeRemoteTags(target);
+        return target ? graph.replace(target) : graph;
+    };
+
+    action.withOption = function(opt) {
```
Comment (jfirebaugh) [inline js/id/actions/merge_remote_changes.js]: Only `safe` is used in practice, right? What's the purpose of `force_local` and `force_remote`?

File: js/id/core/graph.js (hunk 1 above)
Comment (bhousel) [inline js/id/core/graph.js]: It was bothering me that a `Graph` could be initialized with `frozen` left undefined:
```
> g = iD.Graph()
> g.frozen
true
> g2 = iD.Graph(g, true)
> g2.frozen
undefined
```

File: js/id/actions/merge_remote_changes.js [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+            return target.update({nodes: remote.nodes});
+        }
+
+        var o = base.nodes || [],
+            a = local.nodes || [],
+            b = remote.nodes || [],
+            nodes = [],
+            hunks = Diff3.diff3_merge(a, o, b, true);
+
+        for (var i = 0, imax = hunks.length; i !== imax; i++) {
```
Comment (jfirebaugh) [inline js/id/actions/merge_remote_changes.js]: [Don't cache array lengths](http://mrale.ph/blog/2014/12/24/array-length-caching.html).

File: js/id/actions/merge_remote_changes.js [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+            var hunk = hunks[i];
+            if (hunk.ok) {
+                nodes.push.apply(nodes, hunk.ok);
+            }
+            else {
```
Comment (jfirebaugh) [inline js/id/actions/merge_remote_changes.js]: No line break before `else`.

File: js/id/actions/merge_remote_changes.js [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+            return target.update({tags: remote.tags});
+        }
+
+        var keys = _.reject(_.union(_.keys(base.tags), _.keys(remote.tags)), ignoreKey),
+            tags = _.clone(target.tags),
+            changed = false,
+            fail = false;
+
+        function ignoreKey(k) {
+            return k.indexOf('tiger:') === 0 || _.contains(iD.data.discarded, k);
```
Comment (jfirebaugh) [inline js/id/actions/merge_remote_changes.js]: What's the rationale for excluding non-discarded `tiger:*` tags? It looks like the current logic will ignore all remote changes to such keys, which isn't correct.

File: js/id/actions/merge_remote_changes.js [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+        }
+
+        for (var i = 0, imax = keys.length; i !== imax; i++) {
+            var k = keys[i];
+            if (remote.tags[k] !== base.tags[k]) {  // tag modified remotely..
+                if (target.tags[k] && target.tags[k] !== remote.tags[k]) {
```
Comment (jfirebaugh) [inline js/id/actions/merge_remote_changes.js]: In the case of a local delete + remote modify tag conflict, this always resolves in favor of the modification. Is this intended?

File: js/id/actions/merge_remote_changes.js (hunk 5 above)
Comment (bhousel) [inline js/id/actions/merge_remote_changes.js]: I ignored them here because they are ignored by [Entity#hasInterestingTags](https://github.com/openstreetmap/iD/blob/master/js/id/core/entity.js#L111)

File: js/id/core/graph.js (hunk 1 above)
Comment (bhousel) [inline js/id/core/graph.js]: ok

File: js/id/actions/merge_remote_changes.js (hunk 6 above)
Comment (bhousel) [inline js/id/actions/merge_remote_changes.js]: Yes, I did intend to prioritize modifications over deletes.

(This was also why I originally tried to restore locally modified entities when they were deleted on the server.  I still feel like the downside losing a modification is worse than the downside of needing to re-delete something.)

File: js/id/actions/merge_remote_changes.js (hunk 5 above)
Comment (jfirebaugh) [inline js/id/actions/merge_remote_changes.js]: Gotcha. `hasInterestingTags` serves a slightly different purpose though; for conflict resolution I think just `iD.data.discarded` is the right set to ignore.

File: js/id/actions/merge_remote_changes.js (hunk 6 above)
Comment (bhousel) [inline js/id/actions/merge_remote_changes.js]: I'm probably going to change this to raise the conflict rather than silently restoring the tag, since we are probably moving in the direction of giving the user buttons to resolve the conflict anyway.

I think the most common scenario here goes something like:   UserA sees an address that he knows is wrong and deletes it, meanwhile UserB knows the correct street address and enters it. I'm ok with telling UserA about the conflict.

**Comment by bhousel** [comment]
Here's how the conflict resolution ui looks now.. 

screenshot 2015-01-12 22 35 21

Each item can be expanded for more info and to present options for the user to resolve the conflict:

screenshot 2015-01-12 22 32 42

Choosing "Keep Mine" or "Keep Theirs" will redo the merge and update the graph, and transition out the item from the list.

Alternatively, the "Download Changes" button will make generate a OsmChange xml and open it as a `data:` url in a new tab.

**Comment by tristen** [comment]
@bhousel whoa this is awesome work! I'd like to make a couple recommendations for the UI.

- __Download changes__ is a power user utility and I would hesitate to add this as a primary action.  It might be better focused as linked text in the description of the conflict message.
- I'd remove __Try again__. If a user closes or cancels the modal screen, makes new edits and clicks save that is effectively doing the same thing. This eliminates managing any expectation around what that button does and simplifies the choices in the dialog.
- Downplay the Cancel action and make this a secondary action.
- Surface _who_ made the change.
- If feasible, it would be great to break the diff review into separate steps. Here's a wireframe walkthrough:








This provides a few affordances. Rather than providing a link to a node/relation/way id  (Like, _node w264620391_) the visual feature could be highlighted on the map as a user steps through the different changes. It's a better flag of what's been changed. It also allows us to treat these's ids [as links](http://www.openstreetmap.org/way/264620391) in the description text of each issue which can be helpful. Lastly, focusing on one thing (ie. _surface tag changed_) makes it easier to land on a decision of whether to keep your version or the recent saved version remotely.

**Comment by bhousel** [comment]
Thanks @tristen, I like the suggestions!

One thing that you mentioned:
>Rather than providing a link to a node/relation/way id (Like, node w264620391) the visual feature could be highlighted on the map as a user steps through the different changes. It's a better flag of what's been changed. It also allows us to treat these's ids as links in the description text of each issue which can be helpful.

The main reason I didn't do this initially is because the modal box covers up the map, so highlighting features isn't too helpful.  But I think I can put the conflict resolution UI into the sidebar and make this work. Especially if we are stepping the user through each conflict rather than just presenting them all, it could work well.

**Comment by tristen** [comment]
> But I think I can put the conflict resolution UI into the sidebar and make this work.

This sounds great!

**Comment by samanpwbb** [comment]
@bhousel @tristen I'm going to spend a day or two on tightening up the UI here :+1:

**Comment by samanpwbb** [comment]
If there's a conflict when current user changes a feature  that has been deleted, current user is hit with blocking error message and can't save:

screen shot 2015-02-10 at 1 07 31 pm

 Ideally in this situation user is given the option to either keep their changes to the feature or delete the feature.

**Comment by bhousel** [comment]
continued on #2525
